### PR TITLE
Add checkpointing to Ray Tune HPO

### PR DESCRIPTION
**Pull request by krfricke**
Saves checkpoints in Ray Tune HPO, enabling advanced schedulers like PBT.

### CODE REVIEW

**Comment by richardliaw** [comment]
```dif
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index 3470a473..acf1503c 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -544,7 +544,8 @@ class Trainer:
             if trial.should_prune():
                 raise optuna.TrialPruned()
         elif self.hp_search_backend == HPSearchBackend.RAY:
-            self._tune_save_checkpoint()
+            if self.global_step % self.args.save_steps == 0:
+                self._tune_save_checkpoint()
             tune.report(objective=self.objective, **metrics)

     def _tune_save_checkpoint(self):
@@ -911,6 +912,8 @@ class Trainer:
             # search.
             _tb_writer = self.tb_writer
             self.tb_writer = None
+            _model = self.model
+            self.model = None
             # Setup default `resources_per_trial` and `reporter`.
             if "resources_per_trial" not in kwargs and self.args.n_gpu > 0:
                 n_jobs = int(kwargs.pop("n_jobs", 1))
```

This allows us to:

1. Not die when tuning BERT and
2. Not be dominated by saving latency.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -893,11 +912,55 @@ def _objective(trial):
             self.tb_writer = None
             # Setup default `resources_per_trial` and `reporter`.
             if "resources_per_trial" not in kwargs and self.args.n_gpu > 0:
-                kwargs["resources_per_trial"] = {"gpu": self.args.n_gpu}
+                n_jobs = int(kwargs.pop("n_jobs", 1))
```
Comment (richardliaw) [inline src/transformers/trainer.py:915]: Consider documenting this too.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -887,21 +907,70 @@ def _objective(trial):
             best_trial = study.best_trial
             best_run = BestRun(str(best_trial.number), best_trial.value, best_trial.params)
         elif self.hp_search_backend == HPSearchBackend.RAY:
-            # The TensorBoard writer does not pickle so we have to remove it (if it exists) while doing the ray hp
-            # search.
+            # The model and TensorBoard writer do not pickle so we have to remove them (if they exists)
+            # while doing the ray hp search.
             _tb_writer = self.tb_writer
             self.tb_writer = None
+            _model = self.model
```
Comment (sgugger) [inline src/transformers/trainer.py:914]: There is no need to save the old model IMO. Being erased by a new instance of model_init is expected behavior when training.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -887,21 +907,70 @@ def _objective(trial):
...
+            self.model = None
             # Setup default `resources_per_trial` and `reporter`.
             if "resources_per_trial" not in kwargs and self.args.n_gpu > 0:
-                kwargs["resources_per_trial"] = {"gpu": self.args.n_gpu}
+                # `args.n_gpu` is considered the total number of GPUs that will be split
```
Comment (sgugger) [inline src/transformers/trainer.py:918]: This is all starting to be a bit too long for Trainer. Is it possible to factor this in a helper function that would lie in integrations?

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -887,21 +907,70 @@ def _objective(trial):
...
+                        "If this is what you want, do not use {cls}. If you would like to use {cls}, "
+                        "make sure you pass `do_eval=True` and `evaluate_during_training=True` in the "
+                        "Trainer `args`.".format(cls=type(kwargs["scheduler"]).__name__)
+                    )
+
             analysis = tune.run(_objective, config=self.hp_space(None), num_samples=n_trials, **kwargs)
             best_trial = analysis.get_best_trial(metric="objective", mode=direction[:3])
             best_run = BestRun(best_trial.trial_id, best_trial.last_result["objective"], best_trial.config)
             self.tb_writer = _tb_writer
+            self.model = _model
```
Comment (sgugger) [inline src/transformers/trainer.py:973]: As mentioned above, not sure it's worth bothering restoring the old model.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! I just worry the Trainer file is starting to get a bit too long so I think we should factor everything we can from the _objective function in a helper function in the integrations module.

**Comment by krfricke** [comment]
Thanks for your suggestions. I moved the bulk of the hp search code to `integrations`, including the objective, since it depends on the search space. Is this what you had in mind?

**Comment by sgugger** [comment]
Yes. I think we can split the function in two: one for ray, one for optuna and avoid a lot of tests this way to have some cleaner code (with a small duplication in the _objective function). I can do it in a separate PR if you want.

**Comment by krfricke** [comment]
That would be great, thanks!

**Comment by sgugger** [comment]
Merging and will follow up then.

**Pull request merged**